File: macos/Sources/BambuDryCore/Net/Messages.swift
import Foundation

// MARK: - Outbound: ams_filament_drying request

public enum DryCtrlMode: Int, Codable, Sendable {
    case off = 0
    case onTime = 1
    case onHumidity = 2
}

/// Mirrors `device/<serial>/request` envelope for AMS drying.
/// Reference: lagos/src/slic3r/GUI/DeviceCore/DevFilaSystemCtrl.cpp:19-57
public struct DryRequest: Codable, Sendable, Equatable {
    public struct Print: Codable, Sendable, Equatable {
        public let command: String         // "ams_filament_drying"
        public let sequence_id: String
        public let ams_id: Int
        public let mode: Int               // DryCtrlMode raw
        public let filament: String
        public let temp: Int
        public let duration: Int           // hours
        public let humidity: Int           // 0 unless mode == onHumidity
        public let rotate_tray: Bool
        public let cooling_temp: Int
        public let close_power_conflict: Bool
    }
    public let print: Print

    public static func start(amsId: Int,
                             filament: String,
                             tempC: Int,
                             hours: Int,
                             sequenceId: String,
                             rotateTray: Bool = false,
                             coolingTemp: Int = 50,
                             closePowerConflict: Bool = false) -> DryRequest {
        DryRequest(print: Print(
            command: "ams_filament_drying",
            sequence_id: sequenceId,
            ams_id: amsId,
            mode: DryCtrlMode.onTime.rawValue,
            filament: filament,
            temp: tempC,
            duration: hours,
            humidity: 0,
            rotate_tray: rotateTray,
            cooling_temp: coolingTemp,
            close_power_conflict: closePowerConflict
        ))
    }

    public static func stop(amsId: Int, sequenceId: String) -> DryRequest {
        DryRequest(print: Print(
            command: "ams_filament_drying",
            sequence_id: sequenceId,
            ams_id: amsId,
            mode: DryCtrlMode.off.rawValue,
            filament: "",
            temp: 0,
            duration: 0,
            humidity: 0,
            rotate_tray: false,
            cooling_temp: 0,
            close_power_conflict: false
        ))
    }
}

// MARK: - Inbound: device/<serial>/report (AMS subset)
//
// Reference: lagos/src/slic3r/GUI/DeviceCore/DevFilaSystem.cpp:630-700.
// Only the fields we need for auto-dry are decoded; the printer publishes many
// more we ignore.

public struct ReportEnvelope: Decodable, Sendable {
    public let print: PrintReport?
}

public struct PrintReport: Decodable, Sendable {
    public let ams: AmsContainer?
    public let print_type: String?       // "idle" / "running" / etc.
    public let gcode_state: String?      // "PRINTING" / "IDLE" / "PAUSE" / "FINISH"
}

public struct AmsContainer: Decodable, Sendable {
    public let ams: [AmsReport]?
}

public struct AmsReport: Decodable, Sendable {
    public let id: String                 // numeric string, may be 128+ for switchers
    public let humidity: String?          // 0..5 humidity-level (legacy AMS only)
    public let humidity_raw: String?      // "47" → 47%, "-1" → invalid (N3F/N3S)
    public let temp: String?              // float as string
    public let info: String?              // hex bitfield, status decoded from bits
    public let dry_time: Int?             // remaining dry minutes
    public let dry_setting: DrySetting?
    public let dry_sf_reason: [Int]?      // CannotDryReason raw values
    public let tray: [TrayReport]?

    public struct DrySetting: Decodable, Sendable {
        public let dry_filament: String?
        public let dry_temperature: Int?
        public let dry_duration: Int?
    }

    public struct TrayReport: Decodable, Sendable {
        public let id: String
        public let tray_type: String?
        public let tray_sub_brands: String?
    }
}

// MARK: - Snapshot derivation

extension AmsReport {
    /// Decode `info` hex string, extract `count` bits starting at `start`.
    /// Reference: lagos/src/slic3r/GUI/DeviceCore/DevUtil.cpp:7-26
    static func bits(in hexString: String, start: Int, count: Int) -> Int? {
        let trimmed = hexString.trimmingCharacters(in: .whitespacesAndNewlines)
        let cleaned: String
        if trimmed.hasPrefix("0x") || trimmed.hasPrefix("0X") {
            cleaned = String(trimmed.dropFirst(2))
        } else {
            cleaned = trimmed
        }
        guard let value = UInt64(cleaned, radix: 16) else { return nil }
        let mask: UInt64 = (1 << count) - 1
        return Int((value >> start) & mask)
    }

    public func dryStatus() -> DryStatus? {
        guard let info, let bits = Self.bits(in: info, start: 4, count: 4) else { return nil }
        return DryStatus(rawValue: bits)
    }

    /// AMS model is encoded in bits 0..3 of the `info` hex string.
    /// Reference: lagos/src/slic3r/GUI/DeviceCore/DevFilaSystem.cpp:585
    public func amsModel() -> AMSModel? {
        guard let info, let bits = Self.bits(in: info, start: 0, count: 4) else { return nil }
        return AMSModel(rawValue: bits)
    }

    /// Derive an AMSSnapshot, auto-detecting the model from the `info` bitfield.
    /// Returns nil if the model can't be decoded.
    public func snapshot() -> AMSSnapshot? {
        guard let model = amsModel() else { return nil }
        return snapshot(model: model)
    }

    /// Derive an AMSSnapshot for the controller using an explicit model.
    public func snapshot(model: AMSModel) -> AMSSnapshot {
        let amsIdInt = Int(id) ?? -1
        let humidity: Int? = humidity_raw.flatMap { Int($0) }.flatMap { $0 < 0 ? nil : $0 }
        let dryStatus = dryStatus()
        let cannotReasons = (dry_sf_reason ?? []).compactMap { CannotDryReason(rawValue: $0) }
        let dominant = tray?.compactMap { $0.tray_type }
            .filter { !$0.isEmpty }
            .first
        let temp = self.temp.flatMap { Double($0) }
        return AMSSnapshot(
            amsId: amsIdInt,
            model: model,
            humidityPercent: humidity,
            isCurrentlyDrying: dryStatus?.isActive ?? false,
            leftDryTimeMinutes: dry_time ?? 0,
            cannotDryReasons: cannotReasons,
            dominantFilamentType: dominant,
            currentTempC: temp
        )
    }
}

File: windows/src/BambuDry.Core/Net/Messages.cs
using System.Globalization;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace BambuDry.Core.Net;

public enum DryCtrlMode
{
    Off = 0,
    OnTime = 1,
    OnHumidity = 2,
}

// MARK: - Outbound: ams_filament_drying request

/// <summary>
/// Mirrors device/{serial}/request envelope for AMS drying.
/// Reference: lagos/src/slic3r/GUI/DeviceCore/DevFilaSystemCtrl.cpp:19-57
/// </summary>
public sealed record DryRequest([property: JsonPropertyName("print")] DryRequest.PrintBody Print)
{
    public sealed record PrintBody(
        [property: JsonPropertyName("command")]              string Command,
        [property: JsonPropertyName("sequence_id")]          string SequenceId,
        [property: JsonPropertyName("ams_id")]               int    AmsId,
        [property: JsonPropertyName("mode")]                 int    Mode,
        [property: JsonPropertyName("filament")]             string Filament,
        [property: JsonPropertyName("temp")]                 int    Temp,
        [property: JsonPropertyName("duration")]             int    Duration,
        [property: JsonPropertyName("humidity")]             int    Humidity,
        [property: JsonPropertyName("rotate_tray")]          bool   RotateTray,
        [property: JsonPropertyName("cooling_temp")]         int    CoolingTemp,
        [property: JsonPropertyName("close_power_conflict")] bool   ClosePowerConflict);

    public static DryRequest Start(int amsId, string filament, int tempC, int hours, string sequenceId,
                                   bool rotateTray = false, int coolingTemp = 50, bool closePowerConflict = false)
        => new(new PrintBody(
            Command: "ams_filament_drying",
            SequenceId: sequenceId,
            AmsId: amsId,
            Mode: (int)DryCtrlMode.OnTime,
            Filament: filament,
            Temp: tempC,
            Duration: hours,
            Humidity: 0,
            RotateTray: rotateTray,
            CoolingTemp: coolingTemp,
            ClosePowerConflict: closePowerConflict));

    public static DryRequest Stop(int amsId, string sequenceId)
        => new(new PrintBody(
            Command: "ams_filament_drying",
            SequenceId: sequenceId,
            AmsId: amsId,
            Mode: (int)DryCtrlMode.Off,
            Filament: "",
            Temp: 0,
            Duration: 0,
            Humidity: 0,
            RotateTray: false,
            CoolingTemp: 0,
            ClosePowerConflict: false));
}

// MARK: - Inbound: device/{serial}/report (AMS subset)
//
// Reference: lagos/src/slic3r/GUI/DeviceCore/DevFilaSystem.cpp:630-700.
// Only fields needed for auto-dry are decoded; the printer publishes many more.

public sealed class ReportEnvelope
{
    [JsonPropertyName("print")] public PrintReport? Print { get; set; }
}

public sealed class PrintReport
{
    [JsonPropertyName("ams")]         public AmsContainer? Ams { get; set; }
    /// <summary>"idle" / "running" / etc.</summary>
    [JsonPropertyName("print_type")]  public string? PrintType { get; set; }
    /// <summary>"PRINTING" / "IDLE" / "PAUSE" / "FINISH" / "RUNNING"</summary>
    [JsonPropertyName("gcode_state")] public string? GcodeState { get; set; }
}

public sealed class AmsContainer
{
    [JsonPropertyName("ams")] public List<AmsReport>? Ams { get; set; }
}

public sealed class AmsReport
{
    /// <summary>Numeric string; may be 128+ for switchers.</summary>
    [JsonPropertyName("id")]            public string Id { get; set; } = "";
    /// <summary>0..5 humidity-level (legacy AMS only).</summary>
    [JsonPropertyName("humidity")]      public string? Humidity { get; set; }
    /// <summary>"47" → 47%, "-1" → invalid (N3F/N3S).</summary>
    [JsonPropertyName("humidity_raw")]  public string? HumidityRaw { get; set; }
    [JsonPropertyName("temp")]          public string? Temp { get; set; }
    /// <summary>Hex bitfield: bits 0..3 = AMSModel, bits 4..7 = DryStatus.</summary>
    [JsonPropertyName("info")]          public string? Info { get; set; }
    /// <summary>Remaining dry minutes.</summary>
    [JsonPropertyName("dry_time")]      public int? DryTime { get; set; }
    [JsonPropertyName("dry_setting")]   public DrySettingReport? DrySetting { get; set; }
    /// <summary>CannotDryReason raw values.</summary>
    [JsonPropertyName("dry_sf_reason")] public List<int>? DrySfReason { get; set; }
    [JsonPropertyName("tray")]          public List<TrayReport>? Tray { get; set; }

    public sealed class DrySettingReport
    {
        [JsonPropertyName("dry_filament")]    public string? DryFilament { get; set; }
        [JsonPropertyName("dry_temperature")] public int? DryTemperature { get; set; }
        [JsonPropertyName("dry_duration")]    public int? DryDuration { get; set; }
    }

    public sealed class TrayReport
    {
        [JsonPropertyName("id")]              public string Id { get; set; } = "";
        [JsonPropertyName("tray_type")]       public string? TrayType { get; set; }
        [JsonPropertyName("tray_sub_brands")] public string? TraySubBrands { get; set; }
    }

    /// <summary>
    /// Decode <paramref name="hexString"/>, extract <paramref name="count"/> bits starting at <paramref name="start"/>.
    /// Reference: lagos/src/slic3r/GUI/DeviceCore/DevUtil.cpp:7-26
    /// </summary>
    public static int? Bits(string hexString, int start, int count)
    {
        var trimmed = hexString.Trim();
        var cleaned = (trimmed.StartsWith("0x", StringComparison.OrdinalIgnoreCase))
            ? trimmed[2..]
            : trimmed;
        if (!ulong.TryParse(cleaned, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var value))
            return null;
        ulong mask = (1UL << count) - 1;
        return (int)((value >> start) & mask);
    }

    public DryStatus? GetDryStatus()
    {
        if (Info is null) return null;
        var bits = Bits(Info, 4, 4);
        if (bits is null) return null;
        return Enum.IsDefined(typeof(DryStatus), bits.Value) ? (DryStatus)bits.Value : null;
    }

    /// <summary>
    /// AMS model is encoded in bits 0..3 of the <c>info</c> hex string.
    /// Reference: lagos/src/slic3r/GUI/DeviceCore/DevFilaSystem.cpp:585
    /// </summary>
    public AmsModel? GetAmsModel()
    {
        if (Info is null) return null;
        var bits = Bits(Info, 0, 4);
        if (bits is null) return null;
        return Enum.IsDefined(typeof(AmsModel), bits.Value) ? (AmsModel)bits.Value : null;
    }

    /// <summary>Auto-detects model from the <c>info</c> bitfield. Returns null if undecodable.</summary>
    public AmsSnapshot? Snapshot()
    {
        var model = GetAmsModel();
        return model is null ? null : Snapshot(model.Value);
    }

    /// <summary>Build a snapshot using an explicit model.</summary>
    public AmsSnapshot Snapshot(AmsModel model)
    {
        var amsIdInt = int.TryParse(Id, out var v) ? v : -1;

        int? humidity = null;
        if (HumidityRaw is { } hr && int.TryParse(hr, out var hv) && hv >= 0)
            humidity = hv;

        var dryStatus = GetDryStatus();

        var cannotReasons = new List<CannotDryReason>();
        if (DrySfReason is { } reasons)
            foreach (var r in reasons)
                if (Enum.IsDefined(typeof(CannotDryReason), r))
                    cannotReasons.Add((CannotDryReason)r);

        string? dominant = null;
        if (Tray is { } trays)
            foreach (var t in trays)
                if (!string.IsNullOrEmpty(t.TrayType)) { dominant = t.TrayType; break; }

        double? tempC = null;
        if (Temp is { } ts && double.TryParse(ts, NumberStyles.Float, CultureInfo.InvariantCulture, out var tv))
            tempC = tv;

        return new AmsSnapshot(
            AmsId: amsIdInt,
            Model: model,
            HumidityPercent: humidity,
            IsCurrentlyDrying: dryStatus?.IsActive() ?? false,
            LeftDryTimeMinutes: DryTime ?? 0,
            CannotDryReasons: cannotReasons,
            DominantFilamentType: dominant,
            CurrentTempC: tempC);
    }
}

/// <summary>Shared System.Text.Json options used by the transport.</summary>
public static class WireJson
{
    public static readonly JsonSerializerOptions Options = new()
    {
        PropertyNameCaseInsensitive = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.Never,
    };
}
